src/trigger/mod.rs

```diff
@@ -0,0 +1,356 @@
+use crate::{Iden, IntoTableRef, SchemaBuilder, SeaRc, SimpleExpr, TableRef};
+use std::fmt;
+
+mod create;
+mod drop;
+
+pub use create::*;
+pub use drop::*;
+
+pub trait Referencable {
+    fn trigger_ref(&self) -> TriggerRef;
+    fn trigger_name(&self) -> String;
+}
+pub trait Droppable: Referencable {
+    fn drop(&self) -> TriggerDropStatement {
+        TriggerDropStatement::new(self.trigger_ref())
+    }
+}
+
+pub trait Creatable: Referencable {
+    fn create(&self) -> TriggerCreateStatement;
+}
+
+pub trait Configurable {
+    fn configure(
+        &self,
+        table: TableRef,
+        event: TriggerEvent,
+        time: TriggerActionTime,
+    ) -> DefinedTrigger;
+    fn before_insert<T: IntoTableRef>(&self, table: T) -> DefinedTrigger {
+        self.configure(
+            table.into_table_ref(),
+            TriggerEvent::Insert,
+            TriggerActionTime::Before,
+        )
+    }
+    fn after_insert<T: IntoTableRef>(&self, table: T) -> DefinedTrigger {
+        self.configure(
+            table.into_table_ref(),
+            TriggerEvent::Insert,
+            TriggerActionTime::After,
+        )
+    }
+    fn before_update<T: IntoTableRef>(&self, table: T) -> DefinedTrigger {
+        self.configure(
+            table.into_table_ref(),
+            TriggerEvent::Update,
+            TriggerActionTime::Before,
+        )
+    }
+    fn after_update<T: IntoTableRef>(&self, table: T) -> DefinedTrigger {
+        self.configure(
+            table.into_table_ref(),
+            TriggerEvent::Update,
+            TriggerActionTime::After,
+        )
+    }
+    fn before_delete<T: IntoTableRef>(&self, table: T) -> DefinedTrigger {
+        self.configure(
+            table.into_table_ref(),
+            TriggerEvent::Delete,
+            TriggerActionTime::Before,
+        )
+    }
+    fn after_delete<T: IntoTableRef>(&self, table: T) -> DefinedTrigger {
+        self.configure(
+            table.into_table_ref(),
+            TriggerEvent::Delete,
+            TriggerActionTime::After,
+        )
+    }
+}
+
+pub type TriggerAction = SimpleExpr;
+pub type TriggerActions = Vec<TriggerAction>;
+
+#[derive(Default, Debug, Clone)]
+pub struct NamedTrigger {
```

Comment:
How about unifying the two into one with `fn named(...)` and `fn unnamed(...)`?

Reply:
The two structs `UnnamedTrigger` and `NamedTrigger` are distinguished because of slightly different interfaces. Generating trigger names is only needed when a user did not specify an explicit name. Differentiating structs within the Trigger enum allows auto-completion.

The Trigger implementation has two methods `new` and `with_name` that are similar to your suggestion:

```
impl Trigger {
    pub fn new() -> Trigger {
        Trigger::UnnamedTrigger(None, None, None)
    }

    pub fn with_name(name: TriggerRef) -> Trigger {
        Trigger::NamedTrigger(name, None, None, None)
    }
}
```

@HigherOrderLogic do you think we should get rid of the two variants and accept that users might inadvertently use unnamed triggers? Typically I prefer being explicit, but I also see the potential to reduce complexity. Happy to hear your suggestion. 😊

Reply:
> do you think we should get rid of the two variants and accept that users might inadvertently use unnamed triggers

I think that an unified enum would be better for end users DX, however, I disagree that unnamed should be the default.

As I stated in the previous comment, we can implement different functions for each variants, eg:

```rs
impl Trigger {
    pub fn unnamed() -> Trigger {
        Trigger::UnnamedTrigger(None, None, None)
    }

    pub fn named(name: TriggerRef) -> Trigger {
        Trigger::NamedTrigger(name, None, None, None)
    }
}
```